convert2rhel/subscription.py

```diff
@@ -277,28 +288,104 @@ def args(self):
 
         return args
 
+    @property
+    def connection_opts(self):
+        connection_opts = {}
+
+        if self.server_url:
+            url_parts = urllib.parse.urlsplit(self.server_url, allow_fragments=False)
+            connection_opts = {"host": url_parts.hostname}
+            connection_opts = {"port": url_parts.port}
+
+            path = urllib.parse.urlunsplit((None, None, url_parts.path, url_parts.query, url_parts.fragment))
+            connection_opts = {"handler": path}
+
+        connection_opts = dbus.Dictionary(connection_opts, signature="ss")
+        return connection_opts
+
     def __call__(self):
         """
-        Run the subscription-manager command.
-
-        Wrapper for running the subscription-manager command that keeps
-        secrets secure.
+        Use dbus to register the system with subscription-manager.
+
+        Status of dbus on various platforms:
+            * RHEL6:
+                * DBUS is present but may not be installed.
+                * Install the dbus package and run /etc/rc.d/init.d/messagebus start
+                * dbus-python-0.83.0 is available
+            * RHEL7:
+                * dbus looks to be installed and run by default
+                * dbus-python-1.1.9 is available
+            * RHEL8:
+            * RHEL9:
+                * dbus looks to be installed and run by default
+                * python3-dbus-1.2.18 is available
+
+        .. seealso::
+            Documentation for the subscription-manager dbus API:
+            https://www.candlepinproject.org/docs/subscription-manager/dbus_objects.html#registerserver
         """
-        if self.password:
-            loggerinst.debug(
-                "Calling command '%s %s'" % (self.cmd, " ".join(hide_secrets(self.args)))
-            )  # lgtm[py/clear-text-logging-sensitive-data]
-            output, ret_code = utils.run_cmd_in_pty(
-                [self.cmd] + self.args, expect_script=(("[Pp]assword: ", self.password + "\n"),), print_cmd=False
-            )
-        else:
-            # Warning: Currently activation_key can only be specified on the CLI. This is insecure
-            # but there's nothing we can do about it for now. Once subscription-manager issue:
-            # https://issues.redhat.com/browse/ENT-4724 is implemented, we can change both password
-            # and activation_key to use a file-based approach to passing the secrets.
-            output, ret_code = utils.run_subprocess([self.cmd] + self.args, print_cmd=False)
+        # Questions:
+        # * Should we check whether dbus is running?
+        # * If it's not running should we install and run it or should we fall back to
+        #   subscription-managr cli?
+        # * If we install it, do we need to revert that after the upgrade/rollback?
+        # * Should checking for DBUS be done in either system_info or checks.py?
+        # * What should the return value of this code be?
```

Comment:
@abadger, are there still some of the questions unanswered after talking to @ptoscano? I haven't looked at the discussion of you two much so here's how I'd see the answers:

> Should we check whether dbus is running?

Yes.

> If it's not running should we install and run it or should we fall back to subscription-managr cli?

We're already decided to add to spec, so it's installed. If not running, we should start it (no fall back to the sub-man cli).

> If we install it, do we need to revert that after the upgrade/rollback?

If the service was not running and we started it, we should stop the service upon rollback. Let's do that though in a separate PR, after we merge https://github.com/oamg/convert2rhel/pull/516.

> Should checking for DBUS be done in either system_info or checks.py?

I'd say I prefer the checks.py but it's not that important.

> What should the return value of this code be?

I believe you can drop the code in `register_system()` that check the output string (introduced in
https://github.com/oamg/convert2rhel/commit/311b0bd363e3a4b170c9ea981194603e8b4e3858#diff-51e9ff11d39778d3b26778b293fc350430a82e2feed0dab2938da7c0069ffdc6R117). Thus, you don't need to return any string from the `__call__()`.

Regarding the return code - when the registration fails the `Register`/`RegisterWithActivationKeys` services raise an exception, right? We've got the loop for 3 registration attempts and to be able to get to another attempt it seems we should catch the exception. Oh, I now see Pino mentions the same in a comment https://github.com/oamg/convert2rhel/pull/540/files#r923341266.

Reply:
<nod>  I think I've captured all of your (and Pino's) feedback in the checklist in this PR's description.  Let me know if I've missed anything.

I'll remove this comment in the code as it is getting outdated.